README
This repository contains utilities to use extended attribute APIs on different
systems. The utilities are straightforward to use and map closely to their
respective library/system calls.

The utilities included are:
- getxattr - Retrieve an extended attribute and writes its data to stdout.
- listxattr - List extended attributes for a filesystem node.
- removexattr - Remove an extended attribute for a filesystem node.
- setxattr - Set an extended attribute for a filesystem node.


ChangeLog
2023-01-17  Erik Larsson  <catacombae@gmail.com>

	Version 1.0.

	* Initial release supporting macOS, Linux and FreeBSD.


autogen.sh
#!/bin/sh
# Run this to generate configure, Makefile.in's, etc

(autoreconf --version) < /dev/null > /dev/null 2>&1 || {
  (autoconf --version) < /dev/null > /dev/null 2>&1 || {
    echo
    echo "**Error**: You must have the GNU Build System (autoconf, automake, "
    echo "libtool, etc) to update the build system.  Download the appropriate "
    echo "packages for your distribution, or get the source tar balls from "
    echo "ftp://ftp.gnu.org/pub/gnu/."
    exit 1
  }
  echo
  echo "**Error**: Your version of autoconf is too old (you need at least 2.57)"
  echo "to update the build system.  Download the appropriate updated package "
  echo "for your distribution, or get the source tar ball from "
  echo "ftp://ftp.gnu.org/pub/gnu/."
  exit 1
}

echo Running autoreconf --verbose --install --force
autoreconf --verbose --install --force


getxattr.c
/*-
 * getxattr.c - Get the data of one of a filesystem node's extended attributes.
 *
 * Copyright (c) 2023 Erik Larsson
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
#include <fcntl.h>
#include <unistd.h>
#endif
#if defined(__FreeBSD__) || defined(__NetBSD__)
#include <sys/extattr.h>
#endif
#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
#include <sys/stat.h>
#endif
#if defined(__APPLE__) || defined(__DARWIN__) || defined(__linux__)
#include <sys/xattr.h>
#endif

int main(int argc, char **argv)
{
	int ret = (EXIT_FAILURE);
	int argp = 1;
#if defined(__FreeBSD__) || defined(__NetBSD__)
	int namespace = EXTATTR_NAMESPACE_USER;
#endif
	int follow_links = 0;
	const char *path = NULL;;
	const char *attr_name = NULL;
#if defined(__APPLE__) || defined(__DARWIN__)
	const char *attr_offset_string = NULL;
	unsigned long long attr_offset = 0;
#endif
#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	int attrfd = -1;
	struct stat attrstat = { 0 };
#endif
	ssize_t attr_size = 0;
	char *attr_data = NULL;
	ssize_t bytes_read;

	while(argp < argc) {
		if(argv[argp][0] != '-') {
			/* Not an option switch. Move on to the mandatory
			 * arguments. */
			break;
		}
		else if(argv[argp][1] == '-') {
			/* Stop parsing options when '--' is encountered. */
			++argp;
			break;
		}
		else if(argv[argp][1] == 'L') {
			follow_links = 1;
			++argp;
		}
#if defined(__FreeBSD__) || defined(__NetBSD__)
#if defined(EXTATTR_NAMESPACE_EMPTY)
		else if(argv[argp][1] == 'e') {
			namespace = EXTATTR_NAMESPACE_EMPTY;
			++argp;
		}
#endif
		else if(argv[argp][1] == 'u') {
			namespace = EXTATTR_NAMESPACE_USER;
			++argp;
		}
		else if(argv[argp][1] == 's') {
			namespace = EXTATTR_NAMESPACE_SYSTEM;
			++argp;
		}
#endif /* defined(__FreeBSD__) || defined(__NetBSD__) */
	}

	path = (argp < argc) ? argv[argp++] : NULL;
	attr_name = (argp < argc) ? argv[argp++] : NULL;
#if defined(__APPLE__) || defined(__DARWIN__)
	attr_offset_string = (argp < argc) ? argv[argp++] : NULL;
#endif

	if(!path || !attr_name || argp < argc) {
		fprintf(stderr, "usage: getxattr [-L"
#if defined(__FreeBSD__) || defined(__NetBSD__)
#if defined(EXTATTR_NAMESPACE_EMPTY)
			"|-e"
#endif
			"|-u|-s"
#endif
			"] <filename> <attribute name>"
#if defined(__APPLE__) || defined(__DARWIN__)
			" [<attribute offset>]"
#endif
			"\n");
		goto out;
	}

#if defined(__APPLE__) || defined(__DARWIN__)
	if(attr_offset_string) {
		char *endptr = NULL;

		errno = 0;
		attr_offset = strtoull(attr_offset_string, &endptr, 0);
		if(errno || ((*endptr || attr_offset > SIZE_MAX) &&
			(errno = EILSEQ)))
		{
			fprintf(stderr, "Invalid offset: %s\n",
				attr_offset_string);
			goto out;
		}
	}
#endif

#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	if(attr_name[0] == '/') {
		fprintf(stderr, "Invalid attribute name \"%s\" (cannot start "
			"with '/').\n", attr_name);
		goto out;
	}

	attrfd = attropen(
		path,
		attr_name,
		O_RDONLY | (follow_links ? 0 : O_NOFOLLOW));
	if(attrfd == -1) {
		fprintf(stderr, "Error while opening extended attribute \"%s\" "
			"of file \"%s\": %s (%d)\n",
			attr_name,
			path,
			strerror(errno),
			errno);
		goto out;
	}
#endif

#if defined(__APPLE__) || defined(__DARWIN__)
	attr_size = getxattr(
		path,
		attr_name,
		NULL,
		0,
		attr_offset,
		follow_links ? 0 : XATTR_NOFOLLOW);
#elif defined(__linux__)
	attr_size = (follow_links ? getxattr : lgetxattr)(
		path,
		attr_name,
		NULL,
		0);
#elif defined(__FreeBSD__) || defined(__NetBSD__)
	attr_size = (follow_links ? extattr_get_file : extattr_get_link)(
		path,
		namespace,
		attr_name,
		NULL,
		0);
#elif (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	if(fstat(attrfd, &attrstat)) {
		attr_size = -1;
	}
	else {
		attr_size = attrstat.st_size;
	}
#else
#error "Don't know how to handle extended attributes on this platform."
#endif /* defined(__APPLE__) || defined(__DARWIN__) ... */
	if(attr_size == -1) {
		fprintf(stderr, "Error while getting size of extended "
			"attribute for path \"%s\" and attribute name "
			"\"%s\": %s (errno=%d)\n",
			path, attr_name, strerror(errno), errno);
		goto out;
	}

	attr_data = calloc(1, sizeof(char) * (attr_size + 1));
	if(attr_data == NULL) {
		fprintf(stderr, "Error while allocating %zd bytes for data "
			"buffer: %s (errno=%d)\n",
			attr_size, strerror(errno), errno);
		goto out;
	}

#if defined(__APPLE__) || defined(__DARWIN__)
	bytes_read = getxattr(
		path,
		attr_name,
		attr_data,
		attr_size,
		attr_offset,
		follow_links ? 0 : XATTR_NOFOLLOW);
#elif defined(__linux__)
	bytes_read = (follow_links ? getxattr : lgetxattr)(
		path,
		attr_name,
		attr_data,
		attr_size);
#elif defined(__FreeBSD__) || defined(__NetBSD__)
	bytes_read = (follow_links ? extattr_get_file : extattr_get_link)(
		path,
		namespace,
		attr_name,
		attr_data,
		attr_size);
#elif (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	bytes_read = read(
		attrfd,
		attr_data,
		attr_size);
#else
#error "Don't know how to handle extended attributes on this platform."
#endif /* defined(__APPLE__) || defined(__DARWIN__) ... */
	if(bytes_read == -1) {
		fprintf(stderr, "Error while getting extended attribute data "
			"for path \"%s\" and attribute name \"%s\": %s "
			"(errno=%d)\n",
			path, attr_name, strerror(errno), errno);
		goto out;
	}
	else if(bytes_read != attr_size) {
		fprintf(stderr, "Partial read while getting extended attribute "
			"data for path \"%s\" and attribute name \"%s\": "
			"%zd/%zd bytes read\n",
			path, attr_name, bytes_read, attr_size);
		goto out;
	}

	if(fwrite(attr_data, attr_size, 1, stdout) != 1) {
		fprintf(stderr, "Error while writing %zd bytes of extended "
			"attribute data to standard output: %s (errno=%d)\n",
			attr_size, strerror(errno), errno);
		goto out;
	}

	ret = (EXIT_SUCCESS);
out:
	if(attr_data) {
		free(attr_data);
	}
#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	if(attrfd != -1) {
		close(attrfd);
	}
#endif

	return ret;
}


setxattr.c
/*-
 * setxattr.c - Set an extended attribute for a filesystem node.
 *
 * Copyright (c) 2023 Erik Larsson
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <stdint.h>

#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
#include <fcntl.h>
#endif
#include <unistd.h>
#if defined(__FreeBSD__) || defined(__NetBSD__)
#include <sys/extattr.h>
#endif
#if defined(__APPLE__) || defined(__DARWIN__) || defined(__linux__)
#include <sys/xattr.h>
#endif

int main(int argc, char **argv)
{
	int ret = (EXIT_FAILURE);
	int argp = 1;
#if defined(__FreeBSD__) || defined(__NetBSD__)
	int namespace = EXTATTR_NAMESPACE_USER;
#endif
	int follow_links = 0;
	const char *path;
	const char *attr_name = NULL;
	const char *attr_data = NULL;
#if defined(__APPLE__) || defined(__DARWIN__)
	const char *attr_offset_string = NULL;
	unsigned long long attr_offset = 0;
#endif
#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	int nodefd = -1;
	int attrdirfd = -1;
	int attrfd = -1;
	ssize_t attr_bytes_written = 0;
#endif
	char *attr_data_alloc = NULL;
	size_t attr_data_size = 0;

	while(argp < argc) {
		if(argv[argp][0] != '-') {
			/* Not an option switch. Move on to the mandatory
			 * arguments. */
			break;
		}
		else if(argv[argp][1] == '-') {
			/* Stop parsing options when '--' is encountered. */
			++argp;
			break;
		}
		else if(argv[argp][1] == 'L') {
			follow_links = 1;
			++argp;
		}
#if defined(__FreeBSD__) || defined(__NetBSD__)
#ifdef EXTATTR_NAMESPACE_EMPTY
		else if(argv[argp][1] == 'e') {
			namespace = EXTATTR_NAMESPACE_EMPTY;
			++argp;
		}
#endif
		else if(argv[argp][1] == 'u') {
			namespace = EXTATTR_NAMESPACE_USER;
			++argp;
		}
		else if(argv[argp][1] == 's') {
			namespace = EXTATTR_NAMESPACE_SYSTEM;
			++argp;
		}
#endif /* defined(__FreeBSD__) || defined(__NetBSD__) */
	}

	path = (argp < argc) ? argv[argp++] : NULL;
	attr_name = (argp < argc) ? argv[argp++] : NULL;
#if defined(__APPLE__) || defined(__DARWIN__)
	attr_offset_string = (argp < argc) ? argv[argp++] : NULL;
#endif
	attr_data = (argp < argc) ? argv[argp++] : NULL;

	if(!path || !attr_name || argp < argc) {
		fprintf(stderr, "usage: setxattr [-L"
#if defined(__FreeBSD__) || defined(__NetBSD__)
#ifdef EXTATTR_NAMESPACE_EMPTY
			"|-e"
#endif
			"|-u|-s"
#endif /* defined(__FreeBSD__) || defined(__NetBSD__) */
			"] <filename> <attribute name> "
#if defined(__APPLE__) || defined(__DARWIN__)
			"[<attribute offset>] "
#endif
			"[<attribute data>]\n");
		goto out;
	}

#if defined(__APPLE__) || defined(__DARWIN__)
	if(attr_offset_string) {
		char *endptr = NULL;

		errno = 0;
		attr_offset = strtoull(attr_offset_string, &endptr, 0);
		if(errno || ((*endptr || attr_offset > SIZE_MAX) &&
			(errno = EILSEQ)))
		{
			fprintf(stderr, "Invalid offset: %s\n",
				attr_offset_string);
			goto out;
		}
	}
#endif

	if(attr_data) {
		attr_data_size = strlen(attr_data);
	}
	else {
		size_t alloc_size = 4096;

		attr_data = NULL;
		attr_data_size = 0;

		while(1) {
			char *new_attr_data;
			ssize_t bytes_read;

			new_attr_data = realloc(attr_data_alloc, alloc_size);
			if(!new_attr_data) {
				fprintf(stderr, "Error while %sallocating "
					"attribute buffer to %zu bytes: %s "
					"(errno=%d)\n",
					attr_data ? "" : "re", alloc_size,
					strerror(errno), errno);
				goto out;
			}

			attr_data_alloc = new_attr_data;
			attr_data = attr_data_alloc;

			bytes_read = read(STDIN_FILENO, 
				&attr_data_alloc[attr_data_size],
				alloc_size - attr_data_size);
			if(bytes_read < 0) {
				fprintf(stderr, "Error while reading xattr "
					"data from stdin: %s (errno=%d)\n",
					strerror(errno), errno);
				goto out;
			}

			if(!bytes_read) {
				break;
			}

			attr_data_size += bytes_read;
			alloc_size *= 2;
		}

		if(alloc_size != attr_data_size) {
			char *new_attr_data;

			new_attr_data =
				realloc(attr_data_alloc, attr_data_size);
			if(!new_attr_data) {
				fprintf(stderr, "Error while shrinking "
					"attribute buffer from %zu to %zu "
					"bytes: %s (errno=%d)\n",
					alloc_size, attr_data_size,
					strerror(errno), errno);
				goto out;
			}

			attr_data_alloc = new_attr_data;
			attr_data = attr_data_alloc;
		}
	}

#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	if(attr_name[0] == '/') {
		fprintf(stderr, "Invalid attribute name \"%s\" (cannot start "
			"with '/').\n", attr_name);
		goto out;
	}

	nodefd = open(
		path,
		O_RDONLY | (follow_links ? 0 : O_NOFOLLOW));
	if(nodefd == -1) {
		fprintf(stderr, "Error while opening node \"%s\": %s (%d)\n",
			path,
			strerror(errno),
			errno);
		goto out;
	}

	attrdirfd = openat(nodefd, ".", O_RDONLY | O_XATTR);
	if(attrdirfd == -1) {
		fprintf(stderr, "Error while opening \"%s\" node's attribute "
			"directory: %s (%d)\n",
			path,
			strerror(errno),
			errno);
		goto out;
	}
#endif

#if defined(__APPLE__) || defined(__DARWIN__)
	if(setxattr(
		path,
		attr_name,
		attr_data,
		attr_data_size,
		attr_offset,
		follow_links ? 0 : XATTR_NOFOLLOW))
#elif defined(__linux__)
	if((follow_links ? setxattr : lsetxattr)(
		path,
		attr_name,
		attr_data,
		attr_data_size,
		0))
#elif defined(__FreeBSD__) || defined(__NetBSD__)
	if((follow_links ? extattr_set_file : extattr_set_link)(
		path,
		namespace,
		attr_name,
		attr_data,
		attr_data_size) < 0)
#elif (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	/* TODO: Solaris allows whole xattr directory hierarchies under a node.
	 * To support creating attributes in xattr directory hierarchies we must
	 * split any pathname components and create directories for them if they
	 * do not exist. At the moment setting such xattrs will fail. */

	/* If the attribute existed before, then we should remove it. */
	if(unlinkat(attrdirfd, attr_name, 0) && errno != ENOENT) {
		fprintf(stderr, "Error while removing existing extended "
			"attribute  \"%s\" for node \"%s\": %s (%d)\n",
			attr_name, path, strerror(errno), errno);
		goto out;
	}

	attrfd = openat(nodefd, attr_name,
		O_WRONLY | O_CREAT | O_EXCL | O_XATTR, 0777);
	if(attrfd == -1) {
		fprintf(stderr, "Error while creating extended attribute "
			"\"%s\" for node \"%s\": %s (%d)\n",
			attr_name, path, strerror(errno), errno);
		goto out;
	}

	/* Write out the attribute data. */
	attr_bytes_written = write(attrfd, attr_data, attr_data_size);
	if(attr_bytes_written >= 0 && attr_bytes_written != attr_data_size) {
		fprintf(stderr, "Partial write while setting extended "
			"attribute \"%s\" for node \"%s\": %zu / %zu bytes "
			"written\n",
			attr_name, path, attr_bytes_written, attr_data_size);
		goto out;
	}
	else if(attr_bytes_written < 0)
#else
#error "Don't know how to handle extended attributes on this platform."
#endif /* defined(__APPLE__) || defined(__DARWIN__) ... */
	{
		fprintf(stderr, "Failed to set extended attribute: %s "
			"(errno=%d)\n",
			strerror(errno), errno);
		goto out;
	}

	ret = (EXIT_SUCCESS);
out:
#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	if(attrfd != -1) {
		close(attrfd);
	}
#endif

	if(attr_data_alloc) {
		free(attr_data_alloc);
	}

#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	if(attrdirfd != -1) {
		close(attrdirfd);
	}

	if(nodefd != -1) {
		close(nodefd);
	}
#endif

	return ret;
}


listxattr.c
/*-
 * listxattr.c - List the extended attributes of a filesystem node.
 *
 * Copyright (c) 2023 Erik Larsson
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
#include <dirent.h>
#include <fcntl.h>
#include <unistd.h>
#endif
#if defined(__FreeBSD__) || defined(__NetBSD__)
#include <sys/extattr.h>
#endif
#if defined(__APPLE__) || defined(__DARWIN__) || defined(__linux__)
#include <sys/xattr.h>
#endif

int main(int argc, char **argv)
{
#if defined(__FreeBSD__) || defined(__NetBSD__)
	static const int namespaces[] = {
#ifdef EXTATTR_NAMESPACE_EMPTY
		EXTATTR_NAMESPACE_EMPTY,
#endif
		EXTATTR_NAMESPACE_USER,
		EXTATTR_NAMESPACE_SYSTEM,
	};
#endif

	int ret = (EXIT_FAILURE);
	int argp = 1;
#if defined(__FreeBSD__) || defined(__NetBSD__)
	int namespaces_start_index = 0;
	int namespaces_end_index = sizeof(namespaces) / sizeof(namespaces[0]);
	size_t i = 0;
#endif
	int follow_links = 0;
	const char *path = NULL;
#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	int attrdirfd = -1;
#endif
	ssize_t attrlist_size = 0;
	char *attrlist = NULL;

	while(argp < argc) {
		if(argv[argp][0] != '-') {
			/* Not an option switch. Move on to the mandatory
			 * arguments. */
			break;
		}
		else if(argv[argp][1] == '-') {
			/* Stop parsing options when '--' is encountered. */
			++argp;
			break;
		}
		else if(argv[argp][1] == 'L') {
			follow_links = 1;
			++argp;
		}
#if defined(__FreeBSD__) || defined(__NetBSD__)
#ifdef EXTATTR_NAMESPACE_EMPTY
		else if(argv[argp][1] == 'e') {
			namespaces_start_index = 0;
			namespaces_end_index = 1;
			++argp;
		}
#endif
		else if(argv[argp][1] == 'u') {
			namespaces_start_index = 1;
			namespaces_end_index = 2;
			++argp;
		}
		else if(argv[argp][1] == 's') {
			namespaces_start_index = 2;
			namespaces_end_index = 3;
			++argp;
		}
#endif /* defined(__FreeBSD__) || defined(__NetBSD__) */
	}

	path = (argp < argc) ? argv[argp++] : NULL;

	if(!path || (argp < argc)) {
		fprintf(stderr, "usage: listxattr [-L"
#if defined(__FreeBSD__) || defined(__NetBSD__)
#ifdef EXTATTR_NAMESPACE_EMPTY
			"|-e"
#endif
			"|-u|-s"
#endif /* defined(__FreeBSD__) || defined(__NetBSD__) */
			"] <filename>\n");
		goto out;
	}

#if defined(__FreeBSD__) || defined(__NetBSD__)
	for(i = namespaces_start_index; i < namespaces_end_index; ++i)
#else
	do
#endif
	{
		ssize_t bytes_read = -1;
		ssize_t ptr = 0;

#if defined(__APPLE__) || defined(__DARWIN__)
		attrlist_size = listxattr(
			path,
			NULL,
			0,
			follow_links ? 0 : XATTR_NOFOLLOW);
#elif defined(__linux__)
		attrlist_size = (follow_links ? listxattr : llistxattr)(
			path,
			NULL,
			0);
#elif defined(__FreeBSD__) || defined(__NetBSD__)
		attrlist_size =
			(follow_links ? extattr_list_file : extattr_list_link)(
				path,
				namespaces[i],
				NULL,
				0);
#elif (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
		int dup_fd = -1;
		DIR *dirp = NULL;
		struct dirent *de = NULL;
		int err = 0;
		int closedir_res = 0;

		attrdirfd = attropen(
			path,
			".",
			O_RDONLY | (follow_links ? 0 : O_NOFOLLOW));
		if(attrdirfd == -1) {
			fprintf(stderr, "Error while opening attribute "
				"directory of \"%s\": %s (%d)\n",
				path, strerror(errno), errno);
			goto out;
		}

		dup_fd = dup(attrdirfd);
		if(dup_fd == -1 || !(dirp = fdopendir(dup_fd))) {
			fprintf(stderr, "Error while getting attribute "
				"directory handle for \"%s\": %s (%d)\n",
				path, strerror(errno), errno);
			if(dup_fd != -1) {
				close(dup_fd);
			}

			goto out;
		}

		errno = 0;
		while((de = readdir(dirp))) {
			if(de->d_name[0] == '.' && (!de->d_name[1] ||
				(de->d_name[1] == '.' && !de->d_name[2])))
			{
				/* Ignore "." / "..". */
				continue;
			}

			attrlist_size += strlen(de->d_name) + 1;
			errno = 0;
		}

		err = errno;
		closedir_res = closedir(dirp);

		if(err) {
			fprintf(stderr, "Error while reading attribute "
				"directory of \"%s\": %s (%d)\n",
				path, strerror(err), err);
			goto out;
		}
		else if(closedir_res) {
			fprintf(stderr, "Error while closing attribute "
				"directory of \"%s\": %s (%d)\n",
				path, strerror(errno), errno);
			goto out;
		}
#else
#error "Don't know how to handle extended attributes on this platform."
#endif /* defined(__APPLE__) || defined(__DARWIN__) ... */
		if(attrlist_size == 0) {
#ifdef DEBUG
			fprintf(stderr, "INFO: No extended attributes found "
				"for path \"%s\".\n", path);
#endif
			continue;
		}
		else if(attrlist_size == -1) {
#if defined(__FreeBSD__) || defined(__NetBSD__)
			if(errno == EPERM) {
				/* This is normal when a filesystem doesn't
				 * support a namespace. Just move on... */
				continue;
			}
#endif

			fprintf(stderr, "Error while getting size of extended "
				"attribute list "
#if defined(__FreeBSD__) || defined(__NetBSD__)
				"of namespace %d "
#endif
				"for path \"%s\": %s "
				"(errno=%d)\n",
#if defined(__FreeBSD__) || defined(__NetBSD__)
				namespaces[i],
#endif
				path, strerror(errno), errno);
			goto out;
		}

		if(attrlist) {
			free(attrlist);
		}

		attrlist = calloc(1, sizeof(char) * (size_t) attrlist_size);
		if(attrlist == NULL) {
			fprintf(stderr, "Error while allocating %zd bytes for "
				"extended attribute list: %s (errno=%d)\n",
				attrlist_size, strerror(errno), errno);
			goto out;
		}

#if defined(__APPLE__) || defined(__DARWIN__)
		bytes_read = listxattr(
			path,
			attrlist,
			attrlist_size,
			follow_links ? 0 : XATTR_NOFOLLOW);
#elif defined(__linux__)
		bytes_read = (follow_links ? listxattr : llistxattr)(
			path,
			attrlist,
			attrlist_size);
#elif defined(__FreeBSD__) || defined(__NetBSD__)
		bytes_read =
			(follow_links ? extattr_list_file : extattr_list_link)(
				path,
				namespaces[i],
				attrlist,
				attrlist_size);
#elif (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
		{
			int dup_fd = -1;
			DIR *dirp = NULL;
			struct dirent *de = NULL;
			int err = 0;
			int closedir_res = 0;

			if(lseek(attrdirfd, 0, SEEK_SET)) {
				fprintf(stderr, "Error while seeking to start "
					"of directory: %s (%d)\n",
					strerror(errno), errno);
				goto out;
			}

			dup_fd = dup(attrdirfd);
			if(dup_fd == -1 || !(dirp = fdopendir(dup_fd))) {
				fprintf(stderr, "Error while getting attribute "
					"directory handle for \"%s\": %s "
					"(%d)\n",
					path, strerror(errno), errno);
				if(dup_fd != -1) {
					close(dup_fd);
				}

				goto out;
			}

			errno = 0;
			while((de = readdir(dirp))) {
				const size_t name_length = strlen(de->d_name);
				const size_t attrlist_remaining =
					attrlist_size - bytes_read;

				if(de->d_name[0] == '.' && (!de->d_name[1] ||
					(de->d_name[1] == '.' &&
					!de->d_name[2])))
				{
					/* Ignore "." / "..". */
					continue;
				}

				if(attrlist_remaining < name_length + 1) {
					fprintf(stderr, "Not enough space for "
						"all attributes in attribute "
						"list. List may have been "
						"modified behind our backs, "
						"please try again.\n");
					err = EINVAL;
					break;
				}

				strcpy(&attrlist[bytes_read], de->d_name);
				bytes_read += name_length + 1;
				errno = 0;
			}

			if(!de) {
				err = errno;
			}

			closedir_res = closedir(dirp);

			if(err) {
				fprintf(stderr, "Error while reading attribute "
					"directory of \"%s\": %s (%d)\n",
					path, strerror(err), err);
				goto out;
			}
			else if(closedir_res) {
				fprintf(stderr, "Error while closing attribute "
					"directory of \"%s\": %s (%d)\n",
					path, strerror(errno), errno);
				goto out;
			}
		}
#else
#error "Don't know how to handle extended attributes on this platform."
#endif /* defined(__APPLE__) || defined(__DARWIN__) ... */
		if(bytes_read < 0) {
			fprintf(stderr, "Error while reading extended "
				"attribute list "
#if defined(__FreeBSD__) || defined(__NetBSD__)
				"of namespace %d "
#endif
				"for path \"%s\": %s "
				"(errno=%d)\n",
#if defined(__FreeBSD__) || defined(__NetBSD__)
				namespaces[i],
#endif
				path, strerror(errno), errno);
			goto out;
		}
		else if(bytes_read != attrlist_size) {
			fprintf(stderr, "Partial read while reading extended "
				"attribute list for path \"%s\": %zd/%zd bytes "
				"read\n",
				path, bytes_read, attrlist_size);
			goto out;
		}

		while(ptr < attrlist_size) {
#if defined(__FreeBSD__) || defined(__NetBSD__)
			char unknown_namespace_string[] =
				"<namespace -XXXXXXXXXX>";
			const char *namespace_string;
			char *cur = &attrlist[ptr + 1];
			unsigned char cur_len =
				*((unsigned char*) &attrlist[ptr]);
#else
			char *cur = &attrlist[ptr];
			int cur_len = strlen(cur);
#endif

#if defined(__FreeBSD__) || defined(__NetBSD__)
			switch(namespaces[i]) {
#ifdef EXTATTR_NAMESPACE_EMPTY
			case EXTATTR_NAMESPACE_EMPTY:
				namespace_string = "";
				break;
#endif
			case EXTATTR_NAMESPACE_USER:
				namespace_string = "<user>";
				break;
			case EXTATTR_NAMESPACE_SYSTEM:
				namespace_string = "<system>";
				break;
			default:
				snprintf(unknown_namespace_string,
					sizeof(unknown_namespace_string),
					"<namespace %d>", namespaces[i]);
				namespace_string = unknown_namespace_string;
				break;
			}
#endif

			fprintf(stdout,
#if defined(__FreeBSD__) || defined(__NetBSD__)
				"%-*s "
#endif
				"%.*s\n",
#if defined(__FreeBSD__) || defined(__NetBSD__)
				(int) sizeof(unknown_namespace_string) - 1,
				namespace_string,
#endif
				cur_len, cur);

			ptr += cur_len + 1;
		}

		if(ptr != attrlist_size) {
			fprintf(stderr, "WARNING: ptr (%zd) != attrlist_size "
				"(%zd)\n",
				ptr, attrlist_size);
		}
	}
#if !(defined(__FreeBSD__) || defined(__NetBSD__))
	while(0);
#endif

	ret = (EXIT_SUCCESS);
out:
	if(attrlist) {
		free(attrlist);
	}

#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
	if(attrdirfd != -1) {
		close(attrdirfd);
	}
#endif

	return ret;
}
